snippets/varia/heatmaply/heatmaply.Rmd

```diff
@@ -0,0 +1,214 @@
+---
+title: "heatmaply"
+author: "Nicolas Damond"
+date: "03/04/2020"
+output: html_document
+---
+
+# Use of the heatmaply package with SCE objects
+Explains how to generate an interactive heatmap from a `SingleCellExperiment` object
+Includes a `summarizeHeatmap` function that calculates the median counts by channel and 
+by cluster (or any other variable in `colData(sce)`).  
+For more information about heatmaply, see the
+[heatmaply vignette](https://cran.r-project.org/web/packages/heatmaply/vignettes/heatmaply.html) 
+and the original [heatmaply publication](https://doi.org/10.1093/bioinformatics/btx657).
+
+### Requirements
+- Packages: `SingleCellExperiment`, `heatmaply`, `data.table`.  
+- An SCE object with the channels as `rownames(sce)`.
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+knitr::opts_knit$set(root.dir = "./")
+```
+
+## Load the librairies
+```{r librairies, message=FALSE}
+library(SingleCellExperiment)
+library(heatmaply)
+```
+
+## Load example SCE
+Contains a subset of the pancreas dataset
+```{r read-in-data}
+fn.sce <- file.path('..', '..', '..', 'data', 'pancreas_example_sce.rds')
+sce <- readRDS(fn.sce)
+sce
+```
+
+## Function to summarize the data
+Returns median counts by cluster and by channel. The arguments are the following:  
+`sce` = A SingleCellExperiment object.  
+`expr_values` = A string corresponding to an assay in the sce, should be in `names(sce)`.  
+`cluster_by` = Name of the column containing the clusters.  
+`channels`= channels to include, should be in `rownames(sce)`. If `NULL`, all channels will be summarized.
+```{r summarizeHeatmap}
+summarizeHeatmap <- function(sce, expr_values, cluster_by, channels=NULL){
+  require(data.table)
+
+  # Argument checks
+  if(is.null(expr_values)){
+    expr_values <- names(assays(sce))[1]
+    print(paste0("Warning: Assay type not provided, '", expr_values, "' used."))
+  }
+
+  if(is.null(channels[1])){
+    channels <- rownames(sce)
+  }
+
+  if(!all(channels %in% rownames(sce))){
+    stop("Channel names do not correspond to the rownames of the assay")
+  }
+
+  if(is.null(cluster_by)){
+    stop("Cluster column not provided")
+  }
+
+  if(!cluster_by %in% colnames(colData(sce))){
+    stop("The cluster_by argument should correspond to a colData(sce) column")
+  }
+
+  # Convert the data to a melted format
+  dat <- as.data.table(t(assay(sce, expr_values)[channels,]))
+  dat[, id := colnames(sce)]
+  dat[, cluster := colData(sce)[, cluster_by]]
+  dat <- melt.data.table(dat, id.vars=c('id','cluster'), variable.name='channel', value.name=expr_values)
+
+  # Summarize the data
+  dat.summary <- dat[, list(
+    median.val = median(get(expr_values)),
+    mean.val = mean(get(expr_values)),
+    cellspercluster = .N),
+    by = c('channel', 'cluster')]
+
+  # Decast the summarized data and convert to a matrix. Currently only the median values are returned
```

Comment:
You could think of having another argument `FUN` that takes either `mean` or `median` to summarize the counts.